### PR TITLE
Hermite polynomials and transient flow solution

### PR DESCRIPTION
**Pull request by MahdiRoshani**

### CODE REVIEW

File: VascularFlow/Coupling/OneDimensional.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    residual_values = []
+    iteration_indices = []
+    global_inner_counter = 0
+    while time < end_time:
+        time += time_step_size
+        outer_iteration_number += 1
+        inner_iteration_number = 0
+        inner_residual = 1
+        while inner_residual > 10e-08 and inner_iteration_number < 10000:
```
Comment (pastewka) [inline VascularFlow/Coupling/OneDimensional.py]: Please refactor the inner loop into its own function. The residual and number of iterations should become arguments of that function.

File: VascularFlow/Coupling/OneDimensional.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                reynolds_number,
+                strouhal_number,
+                h_star,
+                q_star,
+                q_n,
+                q_n_1,
+            )
+
+            # height calculation
+            h_star = euler_bernoulli_transient(
```
Comment (pastewka) [inline VascularFlow/Coupling/OneDimensional.py]: It's confusing that this is called transient here. The FSI solution should be steady-state I suppose.

File: VascularFlow/Coupling/OneDimensional.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                time_step_size,
+                strouhal_number,
+                inlet_flow_rate,
+                h_star,
+                h_n,
+                h_n_1,
+            )
+
+            # update inner iteration
+            if max(abs(h_new - 1)) < 10e-20:
```
Comment (pastewka) [inline VascularFlow/Coupling/OneDimensional.py]: The tolerance should be an argument of the function, not hard-coded.

File: VascularFlow/Elasticity/Beam.py [hunk 4]
```diff
@@ -22,42 +31,95 @@ def euler_bernoulli(x_n, q_g):
...
+    system_matrix_ll[-2] = 0
 
     # Solve system
-    w_g = np.linalg.solve(system_matrix_gg, q_g)
+    w_g = np.linalg.solve(system_matrix_gg, system_matrix_ll)
+
+    return system_matrix_gg, system_matrix_ll, w_g
+
+
+def euler_bernoulli_transient(x_n, dx_e, num_steps, dt, p, beta, relaxation, H_new):
```
Comment (pastewka) [inline VascularFlow/Elasticity/Beam.py]: This looks like this is doubling a lot of code that is already in the other Euler-Bernoulli function. Try to move that code to a separate utility function so it only appears once. Also please add a documentation string, I don't really understand what this function does. It is actually time-propagating the wave equation on the Euler-Bernoulli beam?

File: VascularFlow/Initialization/InitializeConstants.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+def initialize_constants():
```
Comment (pastewka) [inline VascularFlow/Initialization/InitializeConstants.py:1]: Do not put the simulation parameters in the module. They should go into a separate simulation script.

File: VascularFlow/Initialization/InitializeFlowArrays.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    Parameters:
+        nb_nodes (int): Number of spatial grid points.
+
+    Returns:
+        dict: A dictionary containing initialized arrays.
+    """
+    size = nb_nodes
+    return {
+        "h_n_1": np.ones(size),
```
Comment (pastewka) [inline VascularFlow/Initialization/InitializeFlowArrays.py:16]: Do you actually need this many auxiliary fields? Can you put comments here explaining what each of these fields stores?

File: VascularFlow/Numerics/ArrayFirstDerivative.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import numpy as np
+
+def array_first_derivative(array: np.ndarray, dx_e: float) -> np.ndarray:
```
Comment (pastewka) [inline VascularFlow/Numerics/ArrayFirstDerivative.py]: What does this function do? Please add a documentation string.

It looks like some finite differences scheme to me.

File: VascularFlow/Numerics/Assembly.py [hunk 8]
```diff
@@ -18,3 +17,47 @@ def assemble_system_matrix(element_matrices_enn: ndarray):
         ] += element_matrices_enn[e]
 
     return system_matrix_gg
+
+
+def assemble_system_matrix_2dof(element_matrices_enn: ndarray):
```
Comment (pastewka) [inline VascularFlow/Numerics/Assembly.py]: Does this need to be a separate function? I'd prefer a single function with `dof`  as a parameter. I think it is just a question of changing a couple of indices.

File: VascularFlow/Numerics/Assembly.py [hunk 9]
```diff
@@ -18,3 +17,47 @@ def assemble_system_matrix(element_matrices_enn: ndarray):
...
+    system_matrix_gg = np.zeros(nb_global_nodes)
+
+    for e in range(nb_elements):
+        start = e * (nb_nodes - 2)
+        end = start + nb_nodes
+        system_matrix_gg[start:end] += element_matrices_enn[e].reshape(4,)
+    return system_matrix_gg
+
+
+def assemble_force_matrix_pressure(element_matrices_enn: ndarray):
```
Comment (pastewka) [inline VascularFlow/Numerics/Assembly.py]: Is this the force vector? (Then please rename function to assemble the vector.) How is this specific to pressure? Please also update the documentation string.

File: VascularFlow/Numerics/ElementMatrices.py [hunk 10]
```diff
@@ -14,14 +15,48 @@ def outer_product(x_q: np.array) -> np.array:
 
     return gaussian_quadrature(nb_quad_pts, 0, 1, outer_product)
 
+def dx_matrix(dx: float):
```
Comment (pastewka) [inline VascularFlow/Numerics/ElementMatrices.py]: Why is this necessary? The element matrices should be constructed from the basis functions.

File: VascularFlow/Numerics/ElementMatrices.py [hunk 11]
```diff
@@ -14,14 +15,48 @@ def outer_product(x_q: np.array) -> np.array:
...
+
 
-def second_second(nb_quad_pts: int, basis_function: BasisFunction):
+def second_second(nb_quad_pts: int, dx: int, basis_function: BasisFunction):
     return element_matrix(
         nb_quad_pts, basis_function.second_derivative, basis_function.second_derivative
-    )
+    ) * dx_matrix(dx)
+
+def force_matrix(dx: int):
```
Comment (pastewka) [inline VascularFlow/Numerics/ElementMatrices.py]: Again here, please do not hard code matrices for a specific basis. The idea is to provide a basis set and use the generic quadrature method to compute the integrals.

File: VascularFlow/Numerics/ElementMatrices.py [hunk 12]
```diff
@@ -14,14 +15,48 @@ def outer_product(x_q: np.array) -> np.array:
...
+    return np.array([dx/2, dx**2/12, dx/2, -dx**2/12])
+
+
+def force_matrix_pressure(dx: int):
```
Comment (pastewka) [inline VascularFlow/Numerics/ElementMatrices.py]: Same here.

File: tests/test_assembly.py [hunk 13]
```diff
@@ -25,13 +25,26 @@ def test_assemble_2x2():
...
+    element_matrices_nn = eval_first(1, LinearBasis())
+
+    element_matrices_enn = element_matrices_nn.reshape((1, 2, 2)) * np.ones(3).reshape(
+        (3, 1, 1)
+    )
+
+    system_matrix_gg = assemble_system_matrix_1dof(element_matrices_enn)
+    print(system_matrix_gg.shape)
+    print(system_matrix_gg)
+
```
Comment (pastewka) [inline tests/test_assembly.py:12]: There is some assertion missing here

File: tests/test_beam.py [hunk 14]
```diff
@@ -1,22 +1,58 @@
...
+
+    lhs_matrix = euler_bernoulli(x_n, dx_e, p)[0]
+    rhs_matrix = euler_bernoulli(x_n, dx_e, p)[1]
+    det = np.linalg.det(lhs_matrix)
+
+
+
+    w_g = euler_bernoulli(x_n, dx_e , p)[2]
+    exact = (x_n ** 4 / 24) - (x_n ** 3 / 12) + (x_n ** 2 / 24)
+    print(w_g.shape)
```
Comment (pastewka) [inline tests/test_beam.py]: This is not yet a test. There is an assertion missing.

File: tests/test_element_matrices.py [hunk 15]
```diff
@@ -1,22 +1,46 @@
...
+    print(element_matrix)
 
-    # This is the Laplace matrix for linear elements
-    np.testing.assert_allclose(
-        element_matrix, [[16, -32, 16], [-32, 64, -32], [16, -32, 16]]
-    )
+
+def test_force_matrix():
+    dx = 1
+    print(force_matrix(dx))
```
Comment (pastewka) [inline tests/test_element_matrices.py]: Not a test, assertion missing

File: tests/test_element_matrices.py [hunk 16]
```diff
@@ -1,22 +1,46 @@
...
+
+
+def test_force_matrix_pressure():
+    dx = 1
+    print(force_matrix_pressure(dx))
```
Comment (pastewka) [inline tests/test_element_matrices.py]: Not a test, assertion missing

File: tests/test_element_matrices.py [hunk 17]
```diff
@@ -1,22 +1,46 @@
...
+
+
+def test_dx_matrix_mass():
+    dx = 1
+    print(dx_matrix_mass(dx))
```
Comment (pastewka) [inline tests/test_element_matrices.py]: Not a test, assertion missing

File: tests/test_element_matrices.py [hunk 18]
```diff
@@ -1,22 +1,46 @@
...
+
+
+def test_mass_matrix():
+    basis_function = HermiteBasis()
+    element_matrix = mass_matrix(3, 2, basis_function)
+    print(element_matrix.shape)
+    print(element_matrix)
```
Comment (pastewka) [inline tests/test_element_matrices.py]: Not a test, assertion missing

File: tests/test_flow.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+    # Initialize flow arrays
+    h_n = flow_arrays["h_n"]
+    h_n_1 = flow_arrays["h_n_1"]
+
+    # flow rate calculation
+    qq = flow_rate(
+        x_n, dx_e, dt, strouhal_number, inlet_flow_rate, channel_height, h_n, h_n_1
+    )
+    if plot:
```
Comment (pastewka) [inline tests/test_flow.py:41]: Not a test, assertion missing

File: tests/test_one_dimensional.py [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+        q_n_1,
+        q_n,
+        q_star,
+        q_new,
+        p,
+        p_inner,
+    )
+    print(final_solution[0])
+    print(final_solution[1])
+    print(final_solution[2])
```
Comment (pastewka) [inline tests/test_one_dimensional.py]: Not a test, assertion missing

File: tests/test_pressure.py [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    re = 7.5
+    st = 0.68
+    Hstar = np.ones(len(x_n))
+    Qstar = np.ones(len(x_n))
+    Q_n = np.ones(len(x_n))
+    Q_n1 = np.ones(len(x_n))
+
+    pp = pressure(x_n, dx_e, dt, eps, re, st, Hstar, Qstar, Q_n, Q_n1)
+    print(pp.shape)
+    print(pp)
```
Comment (pastewka) [inline tests/test_pressure.py]: Not a test, assertion missing

File: VascularFlow/Numerics/Assembly.py (hunk 8 above)
Comment (MahdiRoshani) [reply VascularFlow/Numerics/Assembly.py]: Added one general function to create assembly matrix and vectors

File: VascularFlow/Numerics/Assembly.py (hunk 9 above)
Comment (MahdiRoshani) [reply VascularFlow/Numerics/Assembly.py]: Added one general function to create assembly matrix and vectors, and removed all unnecessary functions

File: VascularFlow/Elasticity/Beam.py (hunk 4 above)
Comment (MahdiRoshani) [reply VascularFlow/Elasticity/Beam.py]: Added a function for the clamped boundary condition to avoid repeating lines

File: tests/test_assembly.py (hunk 13 above)
Comment (MahdiRoshani) [reply tests/test_assembly.py:12]: Added assertion

File: VascularFlow/Numerics/ArrayFirstDerivative.py (hunk 7 above)
Comment (MahdiRoshani) [reply VascularFlow/Numerics/ArrayFirstDerivative.py]: Added doc string

File: tests/test_beam.py (hunk 14 above)
Comment (MahdiRoshani) [reply tests/test_beam.py]: Added assertion

File: VascularFlow/Numerics/ElementMatrices.py (hunk 10 above)
Comment (MahdiRoshani) [reply VascularFlow/Numerics/ElementMatrices.py]: Added one general function (load vector) to creta element vector

File: VascularFlow/Numerics/ElementMatrices.py (hunk 11 above)
Comment (MahdiRoshani) [reply VascularFlow/Numerics/ElementMatrices.py]: Added one general function (load vector) to creta element vector

File: VascularFlow/Numerics/ElementMatrices.py (hunk 12 above)
Comment (MahdiRoshani) [reply VascularFlow/Numerics/ElementMatrices.py]: Added one general function (load vector) to creta element vector

File: tests/test_element_matrices.py (hunk 15 above)
Comment (MahdiRoshani) [reply tests/test_element_matrices.py]: Added one test (test_element_matrix_or_vector) for element matrices

File: tests/test_element_matrices.py (hunk 18 above)
Comment (MahdiRoshani) [reply tests/test_element_matrices.py]: Added one test (test_element_matrix_or_vector) for element matrices

File: tests/test_pressure.py (hunk 21 above)
Comment (MahdiRoshani) [reply tests/test_pressure.py]: Added the test (test_pressure)

File: tests/test_flow.py (hunk 19 above)
Comment (MahdiRoshani) [reply tests/test_flow.py:41]: Added assertation

File: VascularFlow/Coupling/OneDimensional.py (hunk 1 above)
Comment (MahdiRoshani) [reply VascularFlow/Coupling/OneDimensional.py]: All done

File: tests/test_one_dimensional.py (hunk 20 above)
Comment (MahdiRoshani) [reply tests/test_one_dimensional.py]: Added a test for a range of physical and numerical parameters.

File: VascularFlow/Coupling/OneDimensional.py (hunk 2 above)
Comment (MahdiRoshani) [reply VascularFlow/Coupling/OneDimensional.py]: I used strong two-way coupling in the simulation. 
It is also possible to have a steady version of the FSI simulation. Which is studied by the paper I attached. 
[s00021-002-8536-9.pdf](https://github.com/user-attachments/files/19588507/s00021-002-8536-9.pdf)

**Comment by MahdiRoshani** [comment]
I am running  test_one_dimensional.py on NEMO2 again. For "inner_it_number=500," it takes too long to approach the next time step!!!

tests/test_one_dimensional.py Time: 0.00250, Inner Iteration: 500, Inner Residual: 3.88945e-04
Time: 0.00500, Inner Iteration: 500, Inner Residual: 1.71895e-04
Time: 0.00750, Inner Iteration: 500, Inner Residual: 1.59746e-04
Time: 0.01000, Inner Iteration: 500, Inner Residual: 4.56094e-04
Time: 0.01250, Inner Iteration: 500, Inner Residual: 1.00939e-03
Time: 0.01500, Inner Iteration: 500, Inner Residual: 1.47809e-03

**Review by pastewka — APPROVED** [review]
